contracts/exchange-forwarder/contracts/src/MixinExchangeWrapper.sol

```diff
@@ -77,183 +80,243 @@ contract MixinExchangeWrapper is
         return fillResults;
     }
 
+    /// @dev Executes a single call of fillOrder according to the wethSellAmount and
+    ///      the amount already sold. Returns false if the transaction would otherwise revert.
+    /// @param order A single order specification.
+    /// @param signature Signature for the given order.
+    /// @param wethSellAmount Total amount of WETH to sell.
+    /// @param wethSpentAmount Amount of WETH already sold.
+    /// @return Amounts filled and fees paid by maker and taker for this order.
+    function _marketSellSingleOrder(
+        LibOrder.Order memory order,
+        bytes memory signature,
+        uint256 wethSellAmount,
+        uint256 wethSpentAmount
+    )
+        internal
+        returns (LibFillResults.FillResults memory singleFillResults)
+    {
+        // The remaining amount of WETH to sell
+        uint256 remainingTakerAssetFillAmount = wethSellAmount.safeSub(wethSpentAmount);
+
+        // Percentage fee
+        if (order.makerAssetData.equals(order.takerFeeAssetData)) {
+            // Attempt to sell the remaining amount of WETH
+            singleFillResults = _fillOrderNoThrow(
+                order,
+                remainingTakerAssetFillAmount,
+                signature
+            );
+        // WETH fee
+        } else if (order.takerFeeAssetData.equals(order.takerAssetData)) {
+            // We will first sell WETH as the takerAsset, then use it to pay the takerFee.
+            // This ensures that we reserve enough to pay the fee.
+            uint256 takerAssetFillAmount = LibMath.getPartialAmountCeil(
+                order.takerAssetAmount,
+                order.takerAssetAmount.safeAdd(order.takerFee),
+                remainingTakerAssetFillAmount
+            );
+
+            singleFillResults = _fillOrderNoThrow(
+                order,
+                takerAssetFillAmount,
+                signature
+            );
+        } else {
+            LibRichErrors.rrevert(LibForwarderRichErrors.UnsupportedFeeError(order.takerFeeAssetData));
+        }
+    }
+
     /// @dev Synchronously executes multiple calls of fillOrder until total amount of WETH has been sold by taker.
     ///      Returns false if the transaction would otherwise revert.
     /// @param orders Array of order specifications.
     /// @param wethSellAmount Desired amount of WETH to sell.
     /// @param signatures Proofs that orders have been signed by makers.
-    /// @return Amounts filled and fees paid by makers and taker.
+    /// @return Amounts of WETH spent and makerAsset acquired by the taker.
```

Comment:
I know we're not very consistent about this, but we should probably start using multiple `@return` directives for functions that have multiple return values.